File: ios/Services/ImageAnalyzer.swift
import Foundation
import CoreImage
import NitroModules

enum ImageAnalyzer {
  private static let noForeground = "No foreground subject detected."

  static func analyze(path: String, options: AnalyzeImageOptions) async throws -> ImageAnalysisResult {
    let segmentOptions = options.removeBackground
    let classifyOptions = options.classify
    let textOptions = options.readText
    let trim = VisionKitOptions.trim(segmentOptions)
    let retainMask = VisionKitOptions.retainMask(segmentOptions)
    let maxPixels = VisionKitOptions.segmentMaxPixels(segmentOptions)
    let maxResults = VisionKitOptions.maxResults(classifyOptions)
    let minConfidence = VisionKitOptions.minConfidence(classifyOptions)
    var classifyRegion = classifyOptions?.region
    var textRegion = textOptions?.region

    let maxLoad = resolveMaxLoad(
      segmentMaxPixels: segmentOptions != nil ? maxPixels : nil,
      classify: classifyOptions != nil,
      readText: textOptions != nil,
    )
    let ciImage = try ImageLoader.loadCIImage(path: path, maxPixels: maxLoad)

    var segmentation: (any HybridSegmentationResultSpec)?
    if segmentOptions != nil {
      do {
        let output = try SubjectSegmenter.segment(ciImage: ciImage, trim: trim, retainMask: retainMask)
        segmentation = HybridSegmentationResult(output: output)
        if classifyRegion == nil, classifyOptions != nil {
          classifyRegion = output.bounds
        }
        if textRegion == nil, textOptions != nil {
          textRegion = output.bounds
        }
      } catch {
        let hasOtherOps = classifyOptions != nil || textOptions != nil
        if !hasOtherOps || !isNoForeground(error) {
          throw error
        }
      }
    }

    var classifications: [Classification]?
    if classifyOptions != nil {
      classifications = try ImageClassifier.classify(
        ciImage: ciImage,
        maxResults: maxResults,
        minConfidence: minConfidence,
        region: classifyRegion,
      )
    }

    var text: (any HybridTextRecognitionResultSpec)?
    if textOptions != nil {
      guard #available(iOS 18.0, *) else {
        throw RuntimeError(VisionAvailability.textRecognitionUnavailableReason)
      }
      let output = try await TextRecognizer.recognize(
        ciImage: ciImage,
        languages: textOptions?.languages,
        recognitionLevel: textOptions?.recognitionLevel,
        region: textRegion,
        minTextHeightFraction: textOptions?.minTextHeightFraction,
        usesLanguageCorrection: textOptions?.usesLanguageCorrection,
        customWords: textOptions?.customWords,
        maxCandidates: textOptions?.maxCandidates,
      )
      text = HybridTextRecognitionResult(output: output)
    }

    return ImageAnalysisResult(
      segmentation: segmentation,
      classifications: classifications,
      text: text,
    )
  }

  private static func isNoForeground(_ error: Error) -> Bool {
    error.localizedDescription.contains(noForeground)
  }

  private static func resolveMaxLoad(
    segmentMaxPixels: Int?,
    classify: Bool,
    readText: Bool,
  ) -> Int {
    var maxLoad = 0
    if let segmentMaxPixels {
      maxLoad = max(maxLoad, segmentMaxPixels)
    }
    if classify {
      maxLoad = max(maxLoad, VisionKitLimits.labelingMaxPixels)
    }
    if readText {
      maxLoad = max(maxLoad, VisionKitLimits.textMaxPixels)
    }
    return maxLoad > 0 ? maxLoad : VisionKitLimits.defaultMaxPixels
  }
}

File: android/src/main/java/com/margelo/nitro/nitrovisionkit/services/ImageAnalyzer.kt
package com.margelo.nitro.nitrovisionkit

import android.content.Context

import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope

internal object ImageAnalyzer {
  private const val NO_FOREGROUND = "No foreground subject detected."

  suspend fun analyze(context: Context, path: String, options: AnalyzeImageOptions): ImageAnalysisResult {
    val segmentOptions = options.removeBackground
    val classifyOptions = options.classify
    val textOptions = options.readText
    val trim = VisionKitOptions.trim(segmentOptions)
    val retainMask = VisionKitOptions.retainMask(segmentOptions)
    val maxPixels = VisionKitOptions.segmentMaxPixels(segmentOptions)
    val maxResults = VisionKitOptions.maxResults(classifyOptions)
    val minConfidence = VisionKitOptions.minConfidence(classifyOptions)
    var classifyRegion = classifyOptions?.region
    var textRegion = textOptions?.region

    val maxLoad = resolveMaxLoad(
      segmentMaxPixels = if (segmentOptions != null) maxPixels else null,
      classify = classifyOptions != null,
      readText = textOptions != null,
    )
    val loaded = ImageLoader.load(context, path, maxLoad)
    try {
      var segmentation: HybridSegmentationResultSpec? = null
      if (segmentOptions != null) {
        try {
          val output = SubjectSegmenter.segment(context, loaded, trim, retainMask)
          segmentation = HybridSegmentationResult(output, ImageLoader.cacheDir(context))
          if (classifyRegion == null && classifyOptions != null) {
            classifyRegion = output.bounds
          }
          if (textRegion == null && textOptions != null) {
            textRegion = output.bounds
          }
        } catch (error: Exception) {
          val hasOtherOps = classifyOptions != null || textOptions != null
          if (!hasOtherOps || error.message?.contains(NO_FOREGROUND) != true) {
            throw error
          }
        }
      }

      var classifications: Array<Classification>? = null
      var text: HybridTextRecognitionResultSpec? = null

      if (classifyOptions != null && textOptions != null) {
        coroutineScope {
          val labelsDeferred = async {
            ImageClassifier.classify(
              loaded,
              maxResults,
              minConfidence,
              classifyRegion,
            ).toTypedArray()
          }
          val textDeferred = async {
            val output = TextRecognizer.recognize(
              context = context,
              bitmap = loaded,
              languages = textOptions.languages,
              region = textRegion,
              minTextHeightFraction = textOptions.minTextHeightFraction,
            )
            HybridTextRecognitionResult(output)
          }
          classifications = labelsDeferred.await()
          text = textDeferred.await()
        }
      } else if (classifyOptions != null) {
        classifications = ImageClassifier.classify(
          loaded,
          maxResults,
          minConfidence,
          classifyRegion,
        ).toTypedArray()
      } else if (textOptions != null) {
        val output = TextRecognizer.recognize(
          context = context,
          bitmap = loaded,
          languages = textOptions.languages,
          region = textRegion,
          minTextHeightFraction = textOptions.minTextHeightFraction,
        )
        text = HybridTextRecognitionResult(output)
      }

      return ImageAnalysisResult(
        segmentation = segmentation,
        classifications = classifications,
        text = text,
      )
    } finally {
      loaded.recycle()
    }
  }

  private fun resolveMaxLoad(
    segmentMaxPixels: Int?,
    classify: Boolean,
    readText: Boolean,
  ): Int {
    var maxLoad = 0
    if (segmentMaxPixels != null) {
      maxLoad = maxOf(maxLoad, segmentMaxPixels)
    }
    if (classify) {
      maxLoad = maxOf(maxLoad, VisionKitLimits.LABELING_MAX_PIXELS)
    }
    if (readText) {
      maxLoad = maxOf(maxLoad, VisionKitLimits.TEXT_MAX_PIXELS)
    }
    return if (maxLoad > 0) maxLoad else VisionKitLimits.DEFAULT_MAX_PIXELS
  }
}
